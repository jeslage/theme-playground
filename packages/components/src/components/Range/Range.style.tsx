import styled from 'styled-components';

const StyledRange = styled.label`
  display: flex;
  align-items: center;
  position: relative;
  width: 100%;

  span {
    margin-right: 10px;
    font-size: 14px;
  }

  input[type='text'] {
    border: none;
    border-bottom: 1px solid transparent;
    outline: none;
    display: inline-block;
    max-width: 30px;
    font-size: inherit;
    padding: 5px 2px 4px;
    font-family: inherit;
    text-align: right;
    color: inherit;
    background: none;

    &:focus {
      border-color: ${({ theme }) => theme.border};
    }
  }

  input[type='range'] {
    appearance: none;
    width: 120px;
    margin: 0;
    background: none;
  }

  input[type='range']:focus {
    outline: none;
  }

  input[type='range']::-webkit-slider-runnable-track {
    width: 100%;
    height: 4px;
    cursor: pointer;
    background: ${({ theme }) => theme.border};
    border-radius: 2px;
  }

  input[type='range']::-moz-range-track {
    width: 100%;
    height: 4px;
    cursor: pointer;
    background: ${({ theme }) => theme.border};
    border-radius: 2px;
  }

  input[type='range']::-ms-track {
    width: 100%;
    height: 4px;
    cursor: pointer;
    background: ${({ theme }) => theme.border};
    border-radius: 2px;
  }

  input[type='range']::-ms-thumb {
    height: 20px;
    width: 20px;
    border-radius: 10px;
    background: ${({ theme }) => theme.text};
    cursor: pointer;
    appearance: none;
    margin-top: -8px;
  }

  input[type='range']::-moz-range-thumb {
    height: 20px;
    width: 20px;
    border-radius: 10px;
    background: ${({ theme }) => theme.text};
    cursor: pointer;
    appearance: none;
    margin-top: -8px;
  }

  input[type='range']::-webkit-slider-thumb {
    height: 20px;
    width: 20px;
    border-radius: 10px;
    background: ${({ theme }) => theme.text};
    cursor: pointer;
    appearance: none;
    margin-top: -8px;
  }
`;

export default StyledRange;
