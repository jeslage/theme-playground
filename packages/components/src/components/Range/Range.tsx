import * as React from 'react';
import { HandleChange } from '../../definitions';

import StyledRange from './Range.style';
import Label from '../Label/Label';

export interface Props {
  iconBefore?: JSX.Element;
  label?: string;
  onChange: (val: number, suffix: string | undefined) => void;
  value: number;
  min?: number;
  max?: number;
  steps?: number;
  suffix?: string | undefined;
  description?: string;
}

const Range: React.FC<Props> = ({
  iconBefore,
  value,
  label,
  onChange,
  min = 0,
  max = 100,
  steps = 1,
  description,
  suffix
}) => {
  const updateValue = (val: number) => {
    if (onChange) onChange(val, suffix);
  };

  const handleChange = (event: HandleChange) => {
    const { value: eventValue, validity } = event.target;

    if (validity.valid) {
      if (eventValue !== '') {
        updateValue(parseFloat(eventValue));
      } else {
        updateValue(0);
      }
    }
  };

  const handleBlur = (event: HandleChange) => {
    const { value: eventValue } = event.target;
    const numberValue = parseFloat(eventValue);

    if (numberValue > max) {
      updateValue(max);
    } else if (numberValue < min) {
      updateValue(min);
    }
  };

  return (
    <StyledRange htmlFor={label}>
      <Label iconBefore={iconBefore} label={label} description={description} />

      <span>
        <input
          type="text"
          pattern="[0-9.]*"
          value={value}
          onChange={handleChange}
          onBlur={handleBlur}
        />
        {suffix}
      </span>

      <input
        type="range"
        step={steps}
        min={min}
        max={max}
        value={value}
        onChange={e => updateValue(parseFloat(e.target.value))}
      />
    </StyledRange>
  );
};

export default Range;
