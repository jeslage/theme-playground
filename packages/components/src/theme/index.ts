export const darkTheme = {
  base: 'dark',
  backgroundDark: '#2f2f2f',
  backgroundLight: '#333',
  border: '#999999',
  text: '#FFFFFF'
};

export const lightTheme = {
  base: 'light',
  backgroundDark: '#f6f9fc',
  backgroundLight: '#fff',
  border: '#666666',
  text: '#333333'
};

export const getUiTheme = theme => {
  if (theme === undefined) return darkTheme;

  if (typeof theme === 'string') {
    if (theme === 'dark') {
      return darkTheme;
    } else if (theme === 'light') {
      return lightTheme;
    }

    throw Error(
      `Theme "${theme}" is not available. Please use "light" or "dark", or provide a custom theme object.`
    );
  }

  if (theme.base && theme.base === 'light') {
    return {
      ...lightTheme,
      ...theme
    };
  }

  return {
    ...darkTheme,
    ...theme
  };
};
