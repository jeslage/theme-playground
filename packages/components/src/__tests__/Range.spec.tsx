import 'jest';
import * as React from 'react';
import { shallow } from 'enzyme';
import Range from '../components/Range/Range';
import StyledRange from '../components/Range/Range.style';
import { darkTheme } from '../theme';

describe('Range', () => {
  it('Range should render correctly', () => {
    const component = shallow(
      <Range
        onChange={value => console.log(value)}
        label="Label"
        description="Testbeschreibung"
        value={1}
        steps={0.1}
        min={0}
        max={2}
        suffix="px"
        iconBefore={
          <svg height="24" viewBox="0 0 24 24" width="24">
            <rect height="2" rx="1" width="12" x="6" y="11" />
          </svg>
        }
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('StyledRange should render correctly', () => {
    const component = shallow(<StyledRange theme={{ colors: darkTheme }} />);

    expect(component).toMatchSnapshot();
  });
});
