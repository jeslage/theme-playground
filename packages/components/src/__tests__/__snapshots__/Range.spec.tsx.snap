// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range Range should render correctly 1`] = `
<styled.label
  htmlFor="Label"
>
  <Label
    description="Testbeschreibung"
    iconBefore={
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <rect
          height="2"
          rx="1"
          width="12"
          x="6"
          y="11"
        />
      </svg>
    }
    label="Label"
  />
  <span>
    <input
      onBlur={[Function]}
      onChange={[Function]}
      pattern="[0-9.]*"
      type="text"
      value={1}
    />
    px
  </span>
  <input
    max={2}
    min={0}
    onChange={[Function]}
    step={0.1}
    type="range"
    value={1}
  />
</styled.label>
`;

exports[`Range StyledRange should render correctly 1`] = `
<label
  className="sc-AxjAm gZeeIQ"
/>
`;
